### ISSUE TITLE
Request for an ifc autosave feature

### ISSUE THREAD

**Issue by sboddy** [IfcOpenShell/IfcOpenShell#5753]
Relative new comers are told not to try and maintain both blender and ifc files. However, a couple of times now I've been caught up in the work - you know that zone where time disappears - and boom, a segfault burns two hours of work. This has happened to me a couple of times in the last few days.

It would be really nice to have an optional autosave with a timer to make a backup ifc file that can be checked for at load time. If the backup is newer than the save file it can ask if you want to load that instead.

**Comment by sboddy** [comment]
Grrrrr!!!! Blender/Bonsai just silently died in the background. Boom! A bunch of tweaks to drawings I now have to repeat. I think this is quite urgently needed.

**Comment by Andrej730** [comment]
> Grrrrr!!!! Blender/Bonsai just silently died in the background. Boom! A bunch of tweaks to drawings I now have to repeat. I think this is quite urgently needed.

Please share the crash report. Are segfaults appearing really often?

**Comment by sboddy** [comment]
In this case the app was in the background, idle - I think I was watching a YouTube video or something. I wasn't running the debugger at the time, so I literally have no idea how to reproduce it. That one was especially annoying because I wasn't doing anything to trigger it. Usually I know what I did to trigger it, and rest assured, I will report those.

I still think the autosave would be useful, at least until Bonsai gets a 1.0 release and a boat-load of users finding all the corner cases. Ideally it eventually becomes redundant because Blender+Bonsai _never_ crashes... 🙏

**Comment by Andrej730** [comment]
> In this case the app was in the background, idle - I think I was watching a YouTube video or something. I wasn't running the debugger at the time, so I literally have no idea how to reproduce it. That one was especially annoying because I wasn't doing anything to trigger it. Usually I know what I did to trigger it, and rest assured, I will report those.

There's typically blender.crash.txt created with the logs.

**Comment by sboddy** [comment]
[blender.crash.txt](https://github.com/user-attachments/files/17784583/blender.crash.txt)

Sure. I think this might be the one - I'm hitting it hard, and so things get a bit fuzzy at 2 am. I doubt it is much use, as there are no debug symbols, so the backtrace is no help. I've already chalked this one down to "S**t happens!".

Am I sensing resistance to the autosave feature, or am just I misreading you trying to help figure out what caused the crash?

**Comment by Andrej730** [comment]
> Sure. I think this might be the one - I'm hitting it hard, and so things get a bit fuzzy at 2 am. I doubt it is much use, as there are no debug symbols, so the backtrace is no help. I've already chalked this one down to "S**t happens!".

Yes, it's not as useful as the tracebacks from Windows that usually have the symbols but still - there's no Python traceback, so it's not directly caused by Bonsai I guess, more of an internal Blender issue. I'll link the traceback here so it will be searchable.

```
# backtrace
/snap/blender/5438/blender() [0xf3dba0]
/snap/blender/5438/blender() [0x84335f]
/lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f21909f5520]
/snap/blender/5438/blender() [0xf64e70]
/snap/blender/5438/blender() [0xf44e79]
/snap/blender/5438/blender() [0xf75262]
/snap/blender/5438/blender() [0xf780a9]
/snap/blender/5438/blender() [0xf7819e]
/snap/blender/5438/blender() [0xf77fae]
/snap/blender/5438/blender() [0x8a3e4f]
/snap/blender/5438/blender() [0x8a3ec6]
/snap/blender/5438/blender() [0x8a3ec6]
/snap/blender/5438/blender() [0xf7827c]
/snap/blender/5438/blender() [0xf7bfea]
/snap/blender/5438/blender() [0xf62336]
/snap/blender/5438/blender() [0xf62522]
/snap/blender/5438/blender() [0xf4605e]
/snap/blender/5438/blender() [0x9f342b]
/snap/blender/5438/blender() [0xf8c9b3]
/snap/blender/5438/blender() [0xf8cae7]
/snap/blender/5438/blender() [0xf95eda]
/snap/blender/5438/blender() [0xf856e8]
/snap/blender/5438/blender() [0x72f480]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f21909dcd90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f21909dce40]
/snap/blender/5438/blender() [0x83ebfe]
```

> Am I sensing resistance to the autosave feature, or am just I misreading you trying to help figure out what caused the crash?

No, I think autosave is a good feature - it probably should be some kind of a popup every n minutes with suggestion to save since saving could take a while.
I think it's just not about the segfaults since there are hardly any, it's more of just Bonsai being alpha software and things might go wrong unexpectedly and it's better to backup from time to time or use ifcgit, even if Bonsai doesn't crash.

**Comment by sboddy** [comment]
I'm using ifcgit too, but that would be even slower than just saving. That's more something I do at the end of the day.

On the autosave behaviour, maybe have the choice. A user configurable timer, and a choice between a reminder (maybe just an INFO that hits the status bar, so it doesn't interrupt workflow) or doing the save directly. My model isn't so big (so far still under 3 MB) so it would be nice to have the option to force autosave. In addition, if the user manually saves it should reset the countdown timer, so users has good, frequent Ctrl-S habits don't get nags/stalls.

Then when autosave itself does the save, it should _not_ overwrite the the main or autosaved file. You ideally want:

- `myfile.ifc`
- `myfile_autosaved.ifc`
- `myfile_autosaving.ifc`

The autosave function saves to `myfile_autosaving.ifc`. when finished, it removes `myfile_autosaved.ifc`, and renames `myfile_autosaving.ifc` to `myfile_autosaved.ifc`. This way there is no write-hole.

If the autosaved file is newer than the main file when opening an ifc, then offer the choice to recover from the autosaved file.
